src/front/glsl/builtins.rs

```diff
@@ -1429,88 +1400,113 @@ impl MacroCall {
+                        SampleLevel::Exact(args[2])
+                    }
+
+                    TextureLevelType::Grad(true) => {
+                        num_args += 2;
                         log::debug!(
                             "Assuming gradients {:?} and {:?} are not greater than 1",
                             args[2],
                             args[3]
                         );
                         SampleLevel::Zero
                     }
-                    false => SampleLevel::Gradient {
-                        x: args[2],
-                        y: args[3],
-                    },
-                };
-                texture_call(ctx, args[0], level, comps, body, meta)
-            }
-            MacroCall::TextureProj => {
-                let level = args
-                    .get(2)
-                    .copied()
-                    .map_or(SampleLevel::Auto, SampleLevel::Bias);
-                let size = match *parser.resolve_type(ctx, args[1], meta)? {
-                    TypeInner::Vector { size, .. } => size,
-                    _ => {
-                        return Err(Error {
-                            kind: ErrorKind::SemanticError("Bad call to textureProj".into()),
-                            meta,
-                        })
+
+                    TextureLevelType::Grad(false) => {
+                        num_args += 2;
+                        SampleLevel::Gradient {
+                            x: args[2],
+                            y: args[3],
+                        }
                     }
                 };
-                let base = args[1];
-                let mut right = ctx.add_expression(
-                    Expression::AccessIndex {
-                        base,
-                        index: size as u32 - 1,
-                    },
-                    SourceMetadata::none(),
-                    body,
-                );
-                let left = if let VectorSize::Bi = size {
-                    ctx.add_expression(
-                        Expression::AccessIndex { base, index: 0 },
-                        SourceMetadata::none(),
-                        body,
-                    )
-                } else {
-                    let size = match size {
-                        VectorSize::Tri => VectorSize::Bi,
-                        _ => VectorSize::Tri,
-                    };
-                    right = ctx.add_expression(
-                        Expression::Splat { size, value: right },
-                        SourceMetadata::none(),
-                        body,
-                    );
-                    ctx.vector_resize(size, base, SourceMetadata::none(), body)
+
+                let texture_offset = match offset {
+                    true => {
+                        let offset_arg = args[num_args];
+                        num_args += 1;
+                        Some(parser.solve_constant(ctx, offset_arg, meta)?)
```

Comment:
Not blocking this, but maybe we could put a `map_err` here and provide a better error message explaining the offset must be a constant expression.

Reply:
I have an ugly way to this, please suggest something better!